lms/djangoapps/completion_api/models.py

```diff
@@ -0,0 +1,249 @@
+    'gp-v2-submission',
+    'gp-v2-peer-selector',
+    'gp-v2-group-selector',
+    'gp-v2-review-question',
+    'gp-v2-peer-assessment',
+    'gp-v2-group-assessment',
+    'gp-v2-static-submissions',
+    'gp-v2-static-grade-rubric',
+    'gp-v2-project-team',
+    'gp-v2-navigator',
+    'gp-v2-navigator-navigation',
+    'gp-v2-navigator-resources',
+    'gp-v2-navigator-submissions',
+    'gp-v2-navigator-ask-ta',
+    'gp-v2-navigator-private-discussion',
+}
+
+
+class CourseCompletionFacade(object):
+    """
+    Facade wrapping progress.models.StudentProgress model.
+    """
+
+    def __init__(self, inner):
+        self._blocks = None
+        self._collected = None
+        self._completable_blocks = None
+        self._inner = inner
+
+    @property
+    def collected(self):
+        """
+        Return the collected block structure for this course
+        """
+        if self._collected is None:
+            self._collected = get_course_in_cache(self.course_key)
+        return self._collected
+
+    @property
+    def blocks(self):
+        """
+        Return all blocks in the course.
+        """
+        if self._blocks is None:
+            course_location = CourseOverview.load_from_module_store(self.course_key).location
+            self._blocks = get_course_blocks(
+                self.user,
+                course_location,
+                collected_block_structure=self.collected,
+            )
+        return self._blocks
+
+    @property
+    def completable_blocks(self):
+        """
+        Return the set of blocks that can be completed that are visible to
+        self.user.
+
+        This method encapsulates the facade's access to the modulestore, making
+        it a useful candidate for mocking.
+        """
+        if self._completable_blocks is None:
+            self._completable_blocks = {
+                block for block in self.blocks
+                if self.blocks.get_xblock_field(block, 'category') not in IGNORE_CATEGORIES
+            }
+        return self._completable_blocks
+
+    @property
+    def user(self):
+        """
+        Return the StudentProgress user
+        """
+        return self._inner.user
+
+    @property
+    def course_key(self):
+        """
+        Return the StudentProgress course_key
+        """
+        return self._inner.course_id
+
+    @property
+    def earned(self):
+        """
+        Return the number of completions earned by the user.
+        """
+
+        return self._inner.completions
+
+    @property
+    def possible(self):
+        """
+        Return the maximum number of completions the user could earn in the
+        course.
+        """
+
+        return float(len(self.completable_blocks))
+
+    @property
+    def percent(self):
+        """
+        Return the percent of the course completed by the user.
+
+        Percent is returned as an integer between 0 and 100.
+        """
+        return int(round(100 * self.earned / self.possible))
```

Comment:
Of course.  I'm thinking if possible is 0, percent should be 100, because the user has completed everything that is possible to complete within the block.